### PR TITLE
perf: prevent duplicate network requests when fetching opportunities data

### LINKED ISSUE #6407: Make `useFetchOpportunities` query based
### Overview

The hook `useFetchOpportunities` currently re-fetches staking opportunity data on every mount, leading to duplicate network requests and performance issues on larger wallets.

### References and additional details

Modify `useFetchOpportunities` such that multiple instances share cache and prevent duplicate network requests where possible.

### Acceptance Criteria

* Multiple instances of `useFetchOpportunities` do not trigger a cascade of duplicate network requests.
* Re-mounting `useFetchOpportunities` does not re-fetch all data.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Refactors `useFetchOpportunities` hook to utilize react-query and rtk query in order to reduce duplicate requests when fetching opportunities data.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #6407

Low risk as this most changes our consumption of existing, unmodified queries.

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

* Check staking opportunities fetch lazily (same as production)
* Check balances on dashboard populate (total, defi tab, rewards tab)
* Check a re-fetch is not triggered when navigating to the dashboard page (unlike production which does)

### CODE REVIEW

**Comment by woodenfurniture** [comment]
* **#6409** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6409?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @woodenfurniture and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/components/StakingVaults/hooks/useFetchOpportunities.tsx
Comment (gomesalexandre) [inline src/components/StakingVaults/hooks/useFetchOpportunities.tsx:1]: Nice catch, we definitely do not to fetch opporunities!

File: src/components/StakingVaults/hooks/useFetchOpportunities.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+      portfolioAssetIds: AssetId[]
+      portfolioAccounts: Record<AccountId, PortfolioAccount>
+    },
+  ] = useMemo(() => {
+    return ['fetchAllOpportunities', { requestedAccountIds, portfolioAssetIds, portfolioAccounts }]
+  }, [portfolioAccounts, portfolioAssetIds, requestedAccountIds])
+
+  const { isLoading } = useQuery({
+    queryKey,
+    queryFn: async ({ queryKey: [_, { requestedAccountIds }] }) => {
```
Comment (gomesalexandre) [inline src/components/StakingVaults/hooks/useFetchOpportunities.tsx:59]: Extract `queryKey` and `queryFn` to `react-queries` as `allOpportunities` and call me here as a function for consistency?

File: src/components/StakingVaults/hooks/useFetchOpportunities.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      if (!requestedAccountIds.length) return
```
Comment (gomesalexandre) [inline src/components/StakingVaults/hooks/useFetchOpportunities.tsx:60]: Do we want this as part of `enabled` instead?

File: src/components/StakingVaults/hooks/useFetchOpportunities.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    {
+      requestedAccountIds: AccountId[]
+      portfolioAssetIds: AssetId[]
+      portfolioAccounts: Record<AccountId, PortfolioAccount>
+    },
+  ] = useMemo(() => {
+    return ['fetchAllOpportunities', { requestedAccountIds, portfolioAssetIds, portfolioAccounts }]
+  }, [portfolioAccounts, portfolioAssetIds, requestedAccountIds])
+
+  const { isLoading } = useQuery({
```
Comment (gomesalexandre) [inline src/components/StakingVaults/hooks/useFetchOpportunities.tsx:57]: @woodenfurniture queries go stale immediately, and have a default garbage collect time of 5mn, do we want to `Infinity` both of these here?

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested against develop after a cache clear on both:

- Opportunities are still not fired on trade page, and lazily on dashboard/earn instead


https://github.com/shapeshift/web/assets/17035424/c89db2a7-664c-4d4a-9928-d43100d5c561

- Navigating back and forth from dashboard/earn and other pages does not trigger a refetch

https://github.com/shapeshift/web/assets/17035424/cf65efa1-a1cc-4f8d-8e54-8e363aecf8db

- Opportunities are still happy 🚫 

image

We never seem to refetch the missing active opportunities for savers - tested by deleting cache and hard refreshing both mutliple times

Other opportunities (e.g read-only, FOXy, UNI-V2, ETH-FOX staking) are also gone in this diff:

image
image
image

On the other hand, tested with Willy's wallet and read-only opportunities are present:

image

To exclude the fact that localhost develop may be rugged against upstream develop, did a sanity check on localhost develop, and confirmed that all opportunities are present same as develop as well.

The issue here seems to be non-deterministic, as more opportunities may sometimes be fetched, sometimes less, the constant being savers ones

**Comment by gomesalexandre** [comment]
@woodenfurniture as found in https://github.com/shapeshift/web/pull/6431, things seem non-determistic in develop already as it sands - feel free to disregard the requested changes after sanity checking this on your side as well!

File: src/components/StakingVaults/hooks/useFetchOpportunities.tsx (hunk 3 above)
Comment (woodenfurniture) [reply src/components/StakingVaults/hooks/useFetchOpportunities.tsx:57]: Absolutely we do, nice catch

File: src/components/StakingVaults/hooks/useFetchOpportunities.tsx (hunk 1 above)
Comment (woodenfurniture) [reply src/components/StakingVaults/hooks/useFetchOpportunities.tsx:59]: I started this and it creates a fair few circular dependencies as the functions in here consume the same react-queries. I've got a branch half ready to resolve this in follow up

**Review by gomesalexandre — APPROVED** [review]
As discussed previously, the spotted issue is non-deterministic and isn't introduced by this PR - stamping as comments have been tackled, and the react-queries extraction has been done in the PR above!

**Pull request merged**